x/confidentialtransfers/types/apply_pending_balance.go

```diff
@@ -0,0 +1,119 @@
+	"errors"
+	"strconv"
+
+	"github.com/sei-protocol/sei-chain/x/confidentialtransfers/utils"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption/elgamal"
+)
+
+// MsgApplyPendingBalance is a message to apply the pending balance to the available balance
+type ApplyPendingBalance struct {
+	Address                        string
+	Denom                          string
+	NewDecryptableAvailableBalance string
+	CurrentPendingBalanceCounter   uint32
+	CurrentAvailableBalance        *elgamal.Ciphertext
+}
+
+// NewApplyPendingBalance creates a new MsgApplyPendingBalance instance
+func NewApplyPendingBalance(
+	privKey ecdsa.PrivateKey,
+	address, denom,
+	currentDecryptableBalance string,
+	currentPendingBalanceCounter uint16,
+	currentAvailableBalance,
+	currentPendingBalanceLo,
+	currentPendingBalanceHi *elgamal.Ciphertext) (*ApplyPendingBalance, error) {
+	aesKey, err := encryption.GetAESKey(privKey, denom)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the current balance from the decryptable balance.
+	currentBalance, err := encryption.DecryptAESGCM(currentDecryptableBalance, aesKey)
+	if err != nil {
+		return nil, err
+	}
+
+	teg := elgamal.NewTwistedElgamal()
+	keyPair, err := teg.KeyGen(privKey, denom)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate the pending balances that we need to add to the available balance.
+	loBalance, err := teg.Decrypt(keyPair.PrivateKey, currentPendingBalanceLo, elgamal.MaxBits32)
+	if err != nil {
+		return nil, err
+	}
+
+	hiBalance, err := teg.DecryptLargeNumber(keyPair.PrivateKey, currentPendingBalanceHi, elgamal.MaxBits48)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the pending balance by combining the lo and hi bits
+	pendingBalance := utils.CombineTransferAmount(uint16(loBalance), uint32(hiBalance))
+
+	// Sum the balances to get the new available balance
+	newDecryptedAvailableBalance := currentBalance + pendingBalance
+
+	// Check for overflow: if the sum is less than one of the operands, an overflow has occurred.
+	if newDecryptedAvailableBalance < currentBalance {
+		return nil, errors.New("addition overflow: total balance exceeds uint64")
+	}
+
+	// Encrypt the new available balance
+	newDecryptableAvailableBalance, err := encryption.EncryptAESGCM(newDecryptedAvailableBalance, aesKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ApplyPendingBalance{
+		Address:                        address,
+		Denom:                          denom,
+		NewDecryptableAvailableBalance: newDecryptableAvailableBalance,
+		CurrentPendingBalanceCounter:   uint32(currentPendingBalanceCounter),
+		CurrentAvailableBalance:        currentAvailableBalance,
+	}, nil
+}
+
+// ApplyPendingBalanceDecrypted is the decrypted version of ApplyPendingBalance
+func (r *ApplyPendingBalance) Decrypt(decryptor *elgamal.TwistedElGamal, privKey ecdsa.PrivateKey, decryptAvailableBalance bool, address string) (*ApplyPendingBalanceDecrypted, error) {
```

Comment:
bool flag signals the method needs to be split in two. e.g `Decrypt` and `DecryptWithAvailableBalance`

Reply:
Tried this but decided not to change in the end - felt a little bit awkward to have 2 methods (Decrypt and DecryptWithAvailableBalance) that do almost the same thing in this specific case, since the only use case here is:
Current:
```
	var result proto.Message
	switch message := sdkmsg.(type) {
	case *types.MsgInitializeAccount:
		result, err = message.Decrypt(decryptor, *privKey, decryptAvailableBalance)
	case *types.MsgWithdraw:
		result, err = message.Decrypt(decryptor, *privKey, decryptAvailableBalance)
	case *types.MsgApplyPendingBalance:
		result, err = message.Decrypt(decryptor, *privKey, decryptAvailableBalance)
	case *types.MsgTransfer:
		result, err = message.Decrypt(decryptor, *privKey, decryptAvailableBalance, address)
	case *types.MsgDeposit:
		result = message
	case *types.MsgCloseAccount:
		result = message
	default:
		return nil, false, nil
```
With separate methods it would become:
```
	var result proto.Message
	switch message := sdkmsg.(type) {
	case *types.MsgInitializeAccount:
                 if decryptAvailableBalance {
                         result, err = message.DecryptWithAvailableBalnace(decryptor, *privKey)
                 } else {
		        result, err = message.Decrypt(decryptor, *privKey)
                 }

	case *types.MsgWithdraw:
		if decryptAvailableBalance {
                         result, err = message.DecryptWithAvailableBalnace(decryptor, *privKey)
                 } else {
		        result, err = message.Decrypt(decryptor, *privKey)
                 }
	case *types.MsgApplyPendingBalance:
                 if decryptAvailableBalance {
                         result, err = message.DecryptWithAvailableBalnace(decryptor, *privKey)
                 } else {
		        result, err = message.Decrypt(decryptor, *privKey)
                 }
	case *types.MsgTransfer:
                 if decryptAvailableBalance {
                         result, err = message.DecryptWithAvailableBalnace(decryptor, *privKey)
                 } else {
		        result, err = message.Decrypt(decryptor, *privKey)
                 }
	case *types.MsgDeposit:
		result = message
	case *types.MsgCloseAccount:
		result = message
	default:
		return nil, false, nil


Point taken for future methods though I think it's a helpful idea